UD verbs on EFA

SeyedMir opened the pull request:
## What
This PR makes the required changes to use UD verbs transport with Amazon EFA NICs.

## Why ?
EFA does not provide certain capabilities that the IB/UD iface assumed to always be there.

## How ?

SeyedMir committed:
UCT/IFACE: Update UD verbs for EFA

swx-jenkins3 commented:
Can one of the admins verify this patch?

yosefe commented:
ok to test

shamisp commented:
@yosefe it would be useful to have CI for this.

yosefe commented on src/uct/ib/base/ib_device.h:
```diff
@@ -210,6 +210,7 @@ typedef struct uct_ib_device {
     int                         numa_node;       /* NUMA node of the device */
     int                         async_events;    /* Whether async events are handled */
     int                         max_zcopy_log_sge; /* Maximum sges log for zcopy am */
+    int                         has_cq_notify;
```
1. uint8_t
2. add comment

yosefe commented on src/uct/ib/ud/base/ud_iface.c:
```diff
@@ -659,6 +673,10 @@ ucs_config_field_t uct_ud_iface_config_table[] = {
      "Max number of receive completions to pick during asynchronous TX poll",
      ucs_offsetof(uct_ud_iface_config_t, rx_async_max_poll), UCS_CONFIG_TYPE_UINT},
 
+    {"TX_CQ_MODERATION", "64",
```
why need to make it configurable?

SeyedMir replied:
With EFA, `IBV_SEND_SIGNALED` must be set in send flags. Otherwise, `ibv_post_send()` will fail. So, CQ moderation must be set to 1 for EFA. But, maybe I should detect the device type and set it to 1 if it is EFA.

yosefe replied:
yes, and let's make sure it's only in verbs code (not in common) so will not affect mlx5

yosefe commented on src/uct/ib/ud/base/ud_iface.c:
```diff
@@ -310,12 +310,25 @@ ucs_status_t uct_ud_iface_complete_init(uct_ud_iface_t *iface)
+            return UCS_OK;
+        }
+
+        status = uct_ib_iface_event_fd_get(&iface->super.super.super, &event_fd);
+        if (status != UCS_OK) {
+            goto err_twheel_cleanup;
```
why need to move uct_ib_iface_event_fd_get() ?

SeyedMir replied:
Because `event_fd` is used only if the condition `(iface->async.event_cb != NULL)` is true.

yosefe commented on src/uct/ib/ud/base/ud_iface.c:
```diff
@@ -310,12 +310,25 @@ ucs_status_t uct_ud_iface_complete_init(uct_ud_iface_t *iface)
-    }
-
     if (iface->async.event_cb != NULL) {
+        uct_iface_attr_t iface_attr;
+        status = uct_ud_iface_query(iface, &iface_attr, 0, 0);
+        if (status != UCS_OK) {
+            goto err_twheel_cleanup;
+        }
+        if (!(iface_attr.cap.event_flags & (UCT_IFACE_FLAG_EVENT_FD |
+                                            UCT_IFACE_FLAG_EVENT_ASYNC_CB))) {
+            ucs_warn("Failed to set callback for async event handling. Device %s "
+                      "does not support any event notification mechanism",
+                       uct_ib_device_name(uct_ib_iface_device(&iface->super)));
+            return UCS_OK;
+        }
```
why ud iface need to query itself?

SeyedMir replied:
I can use `uct_ib_device_has_cq_notify()` instead.

yosefe commented on src/uct/ib/ud/verbs/ud_verbs.c:
```diff
@@ -137,7 +139,7 @@ uct_ud_verbs_ep_send_ctl(uct_ud_ep_t *ud_ep, uct_ud_send_skb_t *skb,
         ucs_assert(!(flags & UCT_UD_IFACE_SEND_CTL_FLAG_INLINE));
     }
-    if (flags & UCT_UD_IFACE_SEND_CTL_FLAG_SOLICITED) {
+    if ((flags & UCT_UD_IFACE_SEND_CTL_FLAG_SOLICITED) && dev->has_cq_notify) {
```
what's wrong if we send solicited if w/o event support? would it cause an error?

SeyedMir replied:
Yes, `ibv_post_send()` will fail.

yosefe replied:
ok

shamisp replied:
@jgunthorpe - I'm just curious if this is standard verbs behavior. thanks !

jgunthorpe replied:
Standard? I'd say no, but it does look right, there is no trace of interrupt handling in EFA for CQs. Polling only.

shamisp replied:
@jgunthorpe If I parse this correctly, you are saying that having CQ interrupt handling is not a strict requirement for verbs providers....

yosefe replied:
According to IBTA spec, event notification (incl. solicited) is mandatory by an HCA:
![image](https://user-images.githubusercontent.com/2255631/108640841-9f3d8600-74a4-11eb-985f-12041e26445b.png)

shamisp replied:
@yosefe EFA is not a IBTA compliant device. AFAIK they have never claimed IBTA or ROCE comparability. I do agree with your point so, if SW IBTA chapter defines the verb as mandatory, it is reasonable to assume that it is implemented and not optional.

yosefe replied:
just to clarify - I think it's not so nice to detect that a mandatory verb is (not) supported by trying to use it, but it's not a blocker for this PR

jgunthorpe replied:
I see Gal asking questions that suggest they might implement it, maybe ask?

SeyedMir replied:
@rajachan any comments on this?

bwbarrett replied:
Pasha is correct; EFA is not an IBTA-compliant device.  When we support a feature, we definitely try to conform to the the IBTA spec as presented by the rdma-core interface, but there are "mandatory" sections we don't support.  While not a public commitment to support a feature, I'm positive we'll eventually have support for completion notifications.  I don't think we'll have that in the time frame NVIDIA is hoping for with this PR.

SeyedMir commented on src/uct/ib/base/ib_md.c:
```diff
@@ -1678,6 +1678,18 @@ void uct_ib_md_close(uct_md_h uct_md)
     ucs_free(md);
 }
 
+static void uct_ib_verbs_md_set_access_flags(uct_ib_md_t *md)
```
But access flags and memory registration are associated with an md, not an iface/tl. For instance, `uct_ib_mem_reg()` gets a `uct_md_h` handle only. So, I'm not sure how I can set different access flags based on iface/tl.

yosefe replied:
We should pass down flags to uct_ib_mem_reg, see uct_md_mem_flags. Should pass UCT_MD_MEM_ACCESS_LOCAL_READ|UCT_MD_MEM_ACCESS_LOCAL_WRITE

SeyedMir replied:
Couldn't figure this out. I can use `uct_md_mem_flags` for memory that is allocated and registered in `uct_iface_mem_alloc()` (commit https://github.com/openucx/ucx/pull/6353/commits/4cebd30a80ad88e7983c708454432acb07e7a861). But, I can't control the flags that are passed to `uct_md_mem_reg()`. For instance, UCP always uses `UCT_MD_MEM_ACCESS_RMA` to register user buffers.
https://github.com/openucx/ucx/blob/7dd283393ad15b75a4e57b98645dc5227fbc4702/src/ucp/core/ucp_request.c#L229
We don't register memory per tl (or communication protocol)

yosefe replied:
what is the blacktrace of failure w/o this change? is it for user buffer registration or for memory pool?
can we change the flags in uct_iface_mp_chunk_alloc -> uct_iface_mem_alloc?

SeyedMir replied:
Without commit https://github.com/openucx/ucx/pull/6353/commits/4cebd30a80ad88e7983c708454432acb07e7a861 the failure happens early on for iface memory pool. But, even with commit https://github.com/openucx/ucx/pull/6353/commits/4cebd30a80ad88e7983c708454432acb07e7a861 user buffer registration will still fail

SeyedMir replied:
Another problem is that some EFA devices support IBV_ACCESS_REMOTE_READ and some don't. So, I cannot set the access flags based on device type either.

yosefe replied:
UD should not really need remote read. we need to narrow down the flags to what is really needed for the datapath operatons.

SeyedMir replied:
Right, I'll see what needs to be changed at UCP layer for this.

SeyedMir replied:
I updated some functions like `ucp_request_send_buffer_reg_lane()` to set the access flags based on the specific communication protocol (commit https://github.com/openucx/ucx/pull/6353/commits/e969ded142ee93b23088ce7b12e0ac32461eab16).
But I'm not sure what we should do for UCP registered memory pool if we want to narrow down the flags.
https://github.com/openucx/ucx/blob/7dd283393ad15b75a4e57b98645dc5227fbc4702/src/ucp/core/ucp_worker.c#L1777

SeyedMir replied:
@yosefe This will disrupt rcache too, because rcache does not know about UCT-level access flags; it only checks for READ/WRITE protection flags. So, if a region is registered with LOCAL_READ | LOCAL_WRITE at some point, and later on we need it registered with REMOTE_READ too, we can't do it because we will get cache hit.
What is the problem with always registering with all the access flags that are supported by the device?

yosefe replied:
the right thing would be to register with flags according to the operation we want to do on the memory
we can pass more flags, if needed, by the user defined arg to rcache_get():
```
 * @param [in]  arg         Custom argument passed down to memory registration
 *                          callback, if a memory registration happens during
 *                          this call.
 * ```

SeyedMir replied:
IIUC, that `arg` is already used to pass down the access flags to the registration callback. Once registered, the access flags are stored in the memory handle. But, `ucs_rcache_get()` cannot use that `arg`, so upon a hit, it cannot check if it is a valid hit with respect to access flags. 
`uct_ib_mem_rcache_reg()` can do that check, and force a re-registration if the access flags of the memory handle do not satisfy those of the `arg`. For that, rcache needs to provide a function for force re-registration.

yosefe replied:
@SeyedMir what if we just replace MD_ACCESS_ALL in https://github.com/openucx/ucx/pull/6353/files#diff-4d9f40ed5ed92d7678c3da565230cac1069f58245a8ebd066af86939f98896efL338 by using "flags" parameter?

yosefe replied:
also, in order to accomplish "register with all supported access flag", I would expect something like just modifying uct_ib_md_access_flags() to check RC transport support and use IBV_LOCAL_xx flags instead of UCT_IB_MEM_ACCESS_FLAGS, and not add anything to uct_base_iface_t

SeyedMir replied:
It won't be enough because that will only solve the problem for the memory that is allocated by an iface. Still, there are other places in UCP where we should decide what access flags should be passed for registration. One example is the registered mpool in UCP worker.

yosefe replied:
AFAIK, ucp mpool should still go through uct_ib_md_access_flags to resolve access flags, isn't this so?

SeyedMir replied:
That's right, and currently that function sets access flags to ACCESS_ALL regardless of what has been passed down.
In the last two commits, I updated that function so it uses the flags that have been passed down to it. But it will still fail since UCP passes ACCESS_RMA/ACCESS_ALL in many places.
That's why I wanted to store in ib md the access flags that are actually supported by the underlying device, so that in that function we use the supported flags (or supported_flags & flags).

yosefe replied:
I think we can check if RC is supported and if not set ib_md::access_flags to LOCAL_READ|LOCAL_WRITE only. 
assuming that if no RC -> no RDMA

SeyedMir replied:
It will work for this PR, but it will fail as soon as we add SRD. Because EFA does not have RC, but it will have RDMA READ which can (and will be) used once we add support for SRD.

yosefe replied:
Is EFA exposing Verbs API?

yosefe replied:
@SeyedMir how would you suggest to check if REMOTE_READ is supported by the device?

SeyedMir replied:
I don't see any other way than actually trying to register memory and see if it fails or not. I could not find anything in device attributes which will tell what flags are supported by a device.

yosefe replied:
how would you know if SRD is supported?
can an SRD-capable device override and add LOCAL_READ flags to the MD?

SeyedMir replied:
All EFA devices will have SRD, but not all of them will have support for REMOTE_READ access flag.

yosefe replied:
check this?
https://github.com/linux-rdma/rdma-core/blob/master/providers/efa/efadv.h#L39

SeyedMir replied:
yes, I think I can query for EFA-specific attributes during device initialization once the device type is detected as EFA.

SeyedMir replied:
@yosefe I updated things accordingly. Can you please take another look at this?

yosefe commented on src/uct/ib/base/ib_device.c:
```diff
@@ -147,6 +147,36 @@ static uct_ib_device_spec_t uct_ib_builtin_device_specs[] = {
   {NULL}
 };
 
+static uint8_t uct_ib_device_has_cq_notify(uct_ib_device_t *dev)
```
here we can return int

yosefe commented on src/uct/ib/ud/accel/ud_mlx5.h:
```diff
@@ -48,8 +48,9 @@ typedef struct {
+    uint16_t tx_moderation = iface->super.config.tx_moderation;
     if ((ce_se & MLX5_WQE_CTRL_CQ_UPDATE) ||
-        (iface->super.tx.unsignaled >= (UCT_UD_TX_MODERATION - 1))) {
+        (iface->super.tx.unsignaled >= (tx_moderation - 1))) {
```
let's not change mlx5

yosefe commented on src/uct/ib/ud/base/ud_iface.c:
```diff
@@ -310,12 +311,19 @@ ucs_status_t uct_ud_iface_complete_init(uct_ud_iface_t *iface)
-
     if (iface->async.event_cb != NULL) {
+        if (!dev->has_cq_notify) {
+            ucs_warn("Failed to set callback for async event handling. Device %s "
```
1. smallcase
2. make it shorter "%s: async event handling is not supported". and print it during device init
3. probably should be ucs_debug

SeyedMir commented on src/uct/ib/base/ib_device.h:
```diff
@@ -441,4 +443,8 @@ static inline ucs_status_t uct_ib_poll_cq(struct ibv_cq *cq, unsigned *count, st
 
+static inline int uct_ib_device_is_efa(const uct_ib_device_t *dev) {
+    return dev->flags & UCT_IB_DEVICE_FLAG_EFA;
+}
```
@yosefe If this is ok, then I can use it instead of `uct_ib_device_has_cq_notify()` to set `has_cq_notify`.

yosefe replied:
IMO, better test it (future proof), than having a hard-coded device id check

SeyedMir committed:
UCT/MD: Add EFA MD and update memory registration access flags

SeyedMir commented:
/azp run

SeyedMir committed:
UCT/IFACE: Add has_inorder_scomp

SeyedMir committed:
UCT/IFACE: Add outstanding map to ud iface

If the IB  device does not guarantee in-order completion
of send requests, the outstanding map is used instead of
the outstanding queue to keep track of send completions.

SeyedMir committed:
UCT/MD: Apply minor fixes

SeyedMir committed:
UCT/MD: Move EFA to separate m4 file

SeyedMir committed:
UCT/MD: Merge the two header files for EFA

SeyedMir committed:
UCT/MD: Extract the common code from verbs md open

SeyedMir committed:
UCT/MD: Reuse verbs key reg/dereg functions for EFA

SeyedMir committed:
UCT/MD: Use uct_ib_efadv_query() to detect EFA device

SeyedMir committed:
UCT/MD: Add efa helper functions

SeyedMir committed:
UCT/MD: Remove UCT_IB_DEVICE_FLAG_EFA

SeyedMir committed:
UCT/MD: Move ibv_open_device to common code and add uct_ib_efadv_check

SeyedMir committed:
UCT/MD: Update helper function for max_rdma_size

SeyedMir committed:
UCT/MD: Build EFA into a separate IB module

SeyedMir committed:
Temporary fix for EFA GDR check

SeyedMir committed:
Set iface comp channel to NULL if device does not support CQ notifications

Otherwise, in case of EFA, a completion channel is created successfully,
and yet causes ibv_create_cq to fail when passed as the argument.

gleon99 commented:
Hi @SeyedMir - we're cleaning up the UCX PR backlog and plan to close this PR in 7 days. Please let us know if you'd like to keep it open.
